### PR TITLE
fix Issue 12368 - std.file.write conflicts with std.stdio.write

### PR DESCRIPTION
**Pull request by CyberShadow**
https://d.puremagic.com/issues/show_bug.cgi?id=12368

### CODE REVIEW

**Comment by andralex** [comment]
Looks like this won't happen. @CyberShadow any code to salvage from here? Any fresh arguments? Or just close?

**Comment by CyberShadow** [comment]
Ehh, this is not exactly blocked by anything. The *proper* way to implement this is blocked by a compiler bug exposed by Phobos doing something unnecessarily weird.

**Comment by CyberShadow** [comment]
Updated... but still blocked - by the same bug, which although has received a PR, was actually fixed incompletely, and thus is reopened.

**Comment by JackStouffer** [comment]
@CyberShadow that bug was re-closed. Time to take another look at this?

**Comment by CyberShadow** [comment]
Yep, on my list

**Comment by quickfur** [comment]
@CyberShadow Can you update the title of this PR? It doesn't reflect a lot of the nice stuff that this PR actually has. For one thing, `lockingBinaryWriter` IMO is a must-have for `std.stdio`. If nothing else goes in, I hope at least that part does!

**Comment by quickfur** [comment]
P.S. Perhaps splitting out `lockingBinaryWriter` to a separate PR would help get it in faster? The more controversial(?) stuff need not hold up the good stuff.

**Comment by CyberShadow** [comment]
No controversy here, just something I need to get around to.

**Comment by CyberShadow** [comment]
OK, fixed up and updated.

**Comment by JackStouffer** [comment]
Proper file writing is beyond me, so I can't really comment on the quality, but I am glad this isn't dead as it has some cool changes :)

**Comment by quickfur** [comment]
`lockingBinaryWriter` is something we ought to have had since a long time ago! (IMO :-P) So the sooner this merges, the better.

File: std/stdio.d [hunk 1]
```diff
@@ -2603,6 +2606,213 @@ See $(LREF byChunk) for an example.
...
+
+    iota(-1, 3, 2.0/size).map!(y =>
+        iota(-1.5, 0.5, 2.0/size).map!(x =>
+            cast(ubyte)(1+
+                recurrence!((a, n) => x + y*1i + a[n-1]^^2)(0+0i)
+                .take(ubyte.max)
+                .countUntil!(z => z.re^^2 + z.im^^2 > 4))
+        )
+    )
+    .copy(stdout.lockingBinaryWriter);
```
Comment (quickfur) [inline std/stdio.d]: :+1: Love this code example.

File: std/stdio.d [hunk 2]
```diff
@@ -2423,8 +2423,13 @@ $(D Range) that locks the file and allows fast writing to it.
     {
     private:
         import std.range.primitives : ElementType, isInfinite, isInputRange;
-        FILE* fps_;          // the shared file handle
-        _iobuf* handle_;     // the unshared version of fps
+        // the shared file handle
+        FILE* fps_;
+
+        // the unshared version of fps
+        @property _iobuf* handle_() @trusted { return cast(_iobuf*)fps_; }
```
Comment (andralex) [inline std/stdio.d]: space after `)`

File: std/stdio.d [hunk 3]
```diff
@@ -2603,6 +2606,213 @@ See $(LREF byChunk) for an example.
...
+                    oldInfo = __fhnd_info[fd];
+                    atomicOp!"&="(__fhnd_info[fd], ~FHND_TEXT);
+                }
+            }
+        }
+
+    public:
+        ~this()
+        {
+            if(fps)
```
Comment (andralex) [inline std/stdio.d]: Space before `(`

Comment (andralex) [inline std/stdio.d]: Let me explain again my pet peeve about this "garden path" code structure (akin to https://en.wikipedia.org/wiki/Garden_path_sentence):

```d
void fun()
{
    if (condition)
    {
        several lines of statements
    }
   // no else, no code after if
}
```

With entering the `if`, the reader makes a mental note to remember that this code is conditional, then starts reading the code inside. At the end of the `if`, the reader figures that conditional code is done and is ready to either go through an `else` clause or some unconditional code that follows. But in fact there is no code following the `if` at all, which means the entire function's body _is_ the conditional. That's a much simpler structure and the reader must drop the mental note, now wasted effort, and reinterpret the function as a do-nothing unless the condition is met. The simpler and superior structure is:

```d
void fun()
{
    if (!condition)
        return;
    several lines of statements
}
```

Now the reader figures immediately that the function is a noop if the condition is met (most often a trivial case that deserves no attention). No more mental note, no more scoping, no more nonlinearity.

File: std/stdio.d [hunk 4]
```diff
@@ -2603,6 +2606,213 @@ See $(LREF byChunk) for an example.
...
+
+        version (Windows)
+        {
+            @disable this(this);
+        }
+        else
+        {
+            this(this)
+            {
+                if(fps)
```
Comment (andralex) [inline std/stdio.d]: space before `(` (your own code does that elsewhere)

File: std/stdio.d (hunk 3 above)
Comment (CyberShadow) [inline std/stdio.d]: I didn't write this code. It is adapted from elsewhere.

File: std/stdio.d [hunk 5]
```diff
@@ -2603,6 +2606,213 @@ See $(LREF byChunk) for an example.
...
+                {
+                    FLOCK(fps);
+                }
+            }
+        }
+
+        void put(T)(auto ref in T value)
+            if (!hasIndirections!T
```
Comment (andralex) [inline std/stdio.d]: `if` flush with `void`

File: std/stdio.d [hunk 6]
```diff
@@ -2603,6 +2606,213 @@ See $(LREF byChunk) for an example.
...
+             && !isInputRange!T)
+        {
+            rawWrite((&value)[0..1]);
+        }
+
+        void put(T)(in T[] array)
+            if (!hasIndirections!T
```
Comment (andralex) [inline std/stdio.d]: ditto

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by andralex** [comment]
Auto-merge toggled off

**Comment by andralex** [comment]
I approve the addition with the nits mentioned. Go for it!

**Comment by CyberShadow** [comment]
Fixed style nits.

File: std/stdio.d (hunk 2 above)
Comment (CyberShadow) [inline std/stdio.d]: [Daniel says no](https://github.com/D-Programming-Language/dlang.org/pull/1191#issuecomment-171106263)

**Comment by JakobOvrum** [comment]
Auto-merge toggled on

**Pull request merged**